File: neuf-log-viewer-api.js
#!/usr/bin/env node

/**
 * NEUF Log Viewer - REST API Server
 * Entry point for REST API mode
 *
 * Usage: node neuf-log-viewer-api.js <folderPath>
 *
 * Provides REST API endpoints for log analysis:
 * - POST /filter_log - Filter logs with pagination (includes filterOptions in response)
 * - POST /export_log - Export filtered logs
 * - POST /preset_suggestions - Get preset filter suggestions
 */

const express = require('express');
const path = require('path');
const fs = require('fs');
const { NEUFLogService } = require('./src/neuf-log-service');

const app = express();
const PORT = process.env.PORT || 3001;

// Parse command-line arguments - folderPath is REQUIRED
const args = process.argv.slice(2);

if (args.length === 0) {
  console.error('❌ Error: Folder path is required');
  console.error('Usage: node neuf-log-viewer-api.js <folderPath>');
  process.exit(0);
}

const folderArg = args[0];
const FOLDER_PATH = path.resolve(folderArg);

if (!fs.existsSync(FOLDER_PATH)) {
  console.error(`❌ Error: Folder path does not exist: ${FOLDER_PATH}`);
  process.exit(0);
}

if (!fs.statSync(FOLDER_PATH).isDirectory()) {
  console.error(`❌ Error: Path is not a directory: ${FOLDER_PATH}`);
  process.exit(0);
}

console.log(`📁 Folder path: ${FOLDER_PATH}`);

// Middleware
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Request logging middleware
app.use((req, res, next) => {
  console.log(`\n📥 ${req.method} ${req.path}`);
  if (req.body && Object.keys(req.body).length > 0) {
    console.log('Body:', JSON.stringify(req.body, null, 2));
  }
  if (req.query && Object.keys(req.query).length > 0) {
    console.log('Query:', JSON.stringify(req.query, null, 2));
  }
  next();
});

// CORS middleware
app.use((req, res, next) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  if (req.method === 'OPTIONS') {
    return res.sendStatus(200);
  }
  next();
});

// Create shared services
const logService = new NEUFLogService(console.log);



/**
 * Parse filters from request body
 * This is API layer responsibility - build filters object from request
 * @param {Object} query - Request body filters object
 * @returns {Object} Parsed and normalized filters
 */
function parseFiltersFromRequest(query) {
  const filters = {
    // Include filters
    filenameInclude: query.filenameInclude || [],
    logLevelInclude: query.logLevelInclude || [],
    threadInclude: query.threadInclude || [],
    deviceInclude: query.deviceInclude || [],
    componentInclude: query.componentInclude || [],
    
    // Exclude filters
    filenameExclude: query.filenameExclude || [],
    logLevelExclude: query.logLevelExclude || [],
    threadExclude: query.threadExclude || [],
    deviceExclude: query.deviceExclude || [],
    componentExclude: query.componentExclude || [],
    
    // Other filters
    timeFrom: query.timeFrom || null,
    timeTo: query.timeTo || null,
    search: query.search || '',
    contextLines: parseInt(query.contextLines) || 0,
    strictContext: query.strictContext === true || query.strictContext === 'true',
    preset: query.preset || ''
  };
  
  // Normalize filters using shared logic from NEUFLogService
  return logService.normalizeFilters(filters);
}

/**
 * Escape CSV field value
 * @param {*} value - Value to escape
 * @returns {string} Escaped CSV value
 */
function escapeCSVValue(value) {
  if (value === null || value === undefined) {
    return '';
  }
  const strValue = String(value);
  if (strValue.includes(',') || strValue.includes('"') || strValue.includes('\n')) {
    return '"' + strValue.replace(/"/g, '""') + '"';
  }
  return strValue;
}




/**
 * POST /preset_suggestions
 * Get preset filter suggestions
 *
 * Response: {
 *   success: true,
 *   suggestions: [
 *     { id, label, description, filters }
 *   ]
 * }
 */
app.post('/preset_suggestions', async (req, res) => {
  try {
    const result = await logService.getPresetSuggestions(FOLDER_PATH);
    res.json(result);
  } catch (error) {
    console.error('❌ Preset suggestions error:', error);
    console.error('Stack trace:', error.stack);
    res.status(500).json({ success: false, error: error.message });
  }
});

/**
 * POST /filter_log
 * Filter logs with a single filters object.
 *
 * Body: {
 *   filters: {
 *     logLevelInclude: ["ERROR"],
 *     deviceInclude: ["DEV001"]
 *   },
 *   page: number (default: 1),
 *   pageSize: number (default: 1000),
 *   inputTable: string (optional) - for applying filters on a specific output table from previous steps
 *   raw: boolean (default: false)
 * }
 *
 * Response: {
 *   success: true,
 *   logs: [...],
 *   total: number,
 *   page: number,
 *   pageSize: number,
 *   totalPages: number,
 *   outputTable: string
 * }
 */
app.post('/filter_log', async (req, res) => {
  try {
    const { filters = {}, steps = [], page = 1, pageSize = 1000, raw = false } = req.body;
    const legacyStep = steps[0] && steps[0].filters ? steps[0].filters : {};
    const requestFilters = Object.keys(filters).length > 0 ? filters : legacyStep;

    // Parse and normalize filters (API layer responsibility)
    const parsedFilters = parseFiltersFromRequest(requestFilters);
    await logService.applyPreset(FOLDER_PATH, parsedFilters);

    const result = await logService.filterLogs(FOLDER_PATH, parsedFilters, { page, pageSize });

    // Fetch filter options from the filtered output table in the same request
    const filterOptionsResult = await logService.getFilterOptions(FOLDER_PATH, result.outputTable);

    if (raw) {
      res.json({ ...result, filterOptions: filterOptionsResult.data });
      return;
    }

    // Format logs at API layer - formatLogEntry now returns object with formattedLog
    const formattedLogs = result.logs.map(log => logService.formatLogEntry(log, "full"));

    res.json({ ...result, logs: formattedLogs, filterOptions: filterOptionsResult.data });

  } catch (error) {
    console.error('❌ Filter chain error:', error);
    console.error('Stack trace:', error.stack);
    res.status(500).json({ success: false, error: error.message });
  }
});

/**
 * POST /export_log
 * Export all logs matching current filters as a downloadable file.
 *
 * Body: {
 *   filters: { ...same as /filter_log },
 *   steps: [{ filters: ... }] (legacy compatibility),
 *   format: "full" | "compact" | "json" (optional, default: "full")
 * }
 */
app.post('/export_log', async (req, res) => {
  try {
    const { filters = {}, steps = [], format = 'full' } = req.body;
    const legacyStep = steps[0] && steps[0].filters ? steps[0].filters : {};
    const requestFilters = Object.keys(filters).length > 0 ? filters : legacyStep;
    const exportFormat = format === 'compact' ? 'compact' : format === 'csv' ? 'csv' : format === 'json' ? 'json' : 'full';

    // ...existing code...
    const parsedFilters = parseFiltersFromRequest(requestFilters);
    await logService.applyPreset(FOLDER_PATH, parsedFilters);

    const exportPageSize = 5000;
    const firstPageResult = await logService.filterLogs(FOLDER_PATH, parsedFilters, { page: 1, pageSize: exportPageSize });

    // Format logs based on export format
    let exportedData;
    if (exportFormat === 'json' || exportFormat === 'csv') {
      // For JSON and CSV, export raw log objects without formattedLog
      exportedData = firstPageResult.logs;
    } else {
      // For 'full' and 'compact', format logs and extract formattedLog
      exportedData = firstPageResult.logs.map(log => logService.formatLogEntry(log, exportFormat));
    }

    for (let page = 2; page <= firstPageResult.totalPages; page++) {
      const pageResult = await logService.filterLogs(FOLDER_PATH, parsedFilters, { page, pageSize: exportPageSize });
      if (exportFormat === 'json' || exportFormat === 'csv') {
        exportedData.push(...pageResult.logs);
      } else {
        const pageData = pageResult.logs.map(log => logService.formatLogEntry(log, exportFormat));
        exportedData.push(...pageData);
      }
    }

    // Generate timestamp in local timezone: YYYY.MM.DD_HH-mm-ss
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const date = String(now.getDate()).padStart(2, '0');
    const hours = String(now.getHours()).padStart(2, '0');
    const minutes = String(now.getMinutes()).padStart(2, '0');
    const seconds = String(now.getSeconds()).padStart(2, '0');
    const timestamp = `${year}.${month}.${date}_${hours}-${minutes}-${seconds}`;

    // Determine file extension and format suffix
    let fileExtension;
    let formatSuffix;

    if (exportFormat === 'json') {
      fileExtension = 'json';
      formatSuffix = 'json';
    } else if (exportFormat === 'csv') {
      fileExtension = 'csv';
      formatSuffix = 'csv';
    } else if (exportFormat === 'compact') {
      fileExtension = 'log';
      formatSuffix = 'compact';
    } else {
      fileExtension = 'log';
      formatSuffix = 'full';
    }

    const filename = `neuf-logs-export-${timestamp}-${formatSuffix}.${fileExtension}`;

    if (exportFormat === 'json') {
      res.setHeader('Content-Type', 'application/json; charset=utf-8');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
      res.send(JSON.stringify(exportedData, null, 2));
    } else if (exportFormat === 'csv') {
      // Convert raw logs to CSV format
      const csvLines = [];

      // CSV header
      if (exportedData.length > 0) {
        const headers = ['filename', 'timestamp', 'log_level', 'thread', 'device', 'component', 'message'];
        csvLines.push(headers.map(h => escapeCSVValue(h)).join(','));

        // CSV rows
        for (const log of exportedData) {
          const row = [
            `(${log.filename})`,
            log.timestamp,
            `[${log.log_level}]`,
            log.thread_name,
            `<${log.device_id}>`,
            `(${log.component_name})`,
            log.message
          ];
          csvLines.push(row.map(v => escapeCSVValue(v)).join(','));
        }
      }

      res.setHeader('Content-Type', 'text/csv; charset=utf-8');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
      res.send(csvLines.join('\n'));
    } else {
      // For 'full' and 'compact' formats, export formattedLog strings
      res.setHeader('Content-Type', 'text/plain; charset=utf-8');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
      const exportedLines = exportedData.map(log => log.formattedLog);
      res.send(exportedLines.join('\n'));
    }

  } catch (error) {
    console.error('❌ Export logs error:', error);
    console.error('Stack trace:', error.stack);
    res.status(500).json({ success: false, error: error.message });
  }
});

/**
 * GET /health
 * Health check endpoint
 */
app.get('/health', (req, res) => {
  const databaseScanned = logService.isDatabaseScanned(FOLDER_PATH);
  
  res.json({
    success: true,
    status: 'healthy',
    version: '1.0.0',
    databaseScanned,
    folderPath: FOLDER_PATH
  });
});

/**
 * Serve static files (for client-side UI)
 */
app.use(express.static(path.join(__dirname, 'public')));

/**
 * Start server
 */
async function main() {
  try {
    // Initialize SQL.js and scan logs
    await logService.initialize();
    
    console.log('');
    console.log('🔍 Scanning logs...');
    const scanResult = await logService.scanLogs(FOLDER_PATH);
    
    if (scanResult.success) {
      if (scanResult.data && !scanResult.data.alreadyScanned) {
        console.log(`✅ Scanned ${scanResult.data.totalLogs} log entries from ${scanResult.data.filesScanned} files`);
      }
    } else {
      console.error('❌ Failed to scan logs:', scanResult.error);
      process.exit(0);
    }
    
    app.listen(PORT, () => {
      console.log('');
      console.log('╔════════════════════════════════════════╗');
      console.log('║   🚀 NEUF Log Viewer API v1.0         ║');
      console.log('╚════════════════════════════════════════╝');
      console.log('');
      console.log(`🌐 API Server running at http://localhost:${PORT}`);
      console.log(`📊 Web UI available at http://localhost:${PORT}`);
      console.log('');
      console.log(`📁 Folder path: ${FOLDER_PATH}`);
      console.log('');
      console.log('📋 Available endpoints:');
      console.log('  POST   /filter_log           - Filter logs (includes filterOptions)');
      console.log('  POST   /export_log           - Export filtered logs');
      console.log('  POST   /preset_suggestions   - Get preset filter suggestions');
      console.log('  GET    /health               - Health check');
      console.log('');
      console.log('Press Ctrl+C to stop the server');
      console.log('');
    });
    
  } catch (error) {
    console.error('');
    console.error('❌ Failed to start server:');
    console.error(error.message || error);
    process.exit(0);
  }
}

main();

File: python-port/neuf_log_viewer_api.py
#!/usr/bin/env python3
"""
NEUF Log Viewer — REST API Server (Python/FastAPI port of neuf-log-viewer-api.js)

Usage:
  python neuf_log_viewer_api.py <folderPath>

Provides REST API endpoints for log analysis:
  POST /filter_log          - Filter logs with pagination (includes filterOptions)
  POST /export_log          - Export filtered logs
  POST /preset_suggestions  - Get preset filter suggestions
  GET  /health              - Health check
"""

import os
import sys
from datetime import datetime
from typing import Any, Dict, List, Optional

import uvicorn
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse, Response
from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Bootstrap: make python-port/src importable
# ---------------------------------------------------------------------------
_HERE = os.path.dirname(os.path.abspath(__file__))
if _HERE not in sys.path:
    sys.path.insert(0, _HERE)

from src.neuf_log_service import NEUFLogService  # noqa: E402


# ---------------------------------------------------------------------------
# Request / response schemas
# ---------------------------------------------------------------------------

class FilterLogRequest(BaseModel):
    filters: Optional[Dict[str, Any]] = None
    steps: Optional[List[Dict[str, Any]]] = None
    page: int = 1
    pageSize: int = 1000
    raw: bool = False


class ExportLogRequest(BaseModel):
    filters: Optional[Dict[str, Any]] = None
    steps: Optional[List[Dict[str, Any]]] = None
    format: str = 'full'


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def parse_filters_from_request(query: dict, log_service) -> dict:
    """
    Build a normalised filters dict from a raw request body dict.
    This is the API layer's responsibility (mirrors parseFiltersFromRequest in JS).
    """
    has_context_lines = 'contextLines' in query
    has_strict_context = 'strictContext' in query

    raw_context = query.get('contextLines')
    if has_context_lines:
        try:
            context_lines = int(raw_context) if raw_context is not None else 0
        except (ValueError, TypeError):
            context_lines = 0
    else:
        context_lines = None  # sentinel: not present

    raw_strict = query.get('strictContext')
    if has_strict_context:
        strict_context = raw_strict is True or str(raw_strict).lower() == 'true'
    else:
        strict_context = None  # sentinel: not present

    filters = {
        'filenameInclude': query.get('filenameInclude') or [],
        'logLevelInclude': query.get('logLevelInclude') or [],
        'threadInclude': query.get('threadInclude') or [],
        'deviceInclude': query.get('deviceInclude') or [],
        'componentInclude': query.get('componentInclude') or [],

        'filenameExclude': query.get('filenameExclude') or [],
        'logLevelExclude': query.get('logLevelExclude') or [],
        'threadExclude': query.get('threadExclude') or [],
        'deviceExclude': query.get('deviceExclude') or [],
        'componentExclude': query.get('componentExclude') or [],

        'timeFrom': query.get('timeFrom') or None,
        'timeTo': query.get('timeTo') or None,
        'search': query.get('search') or '',
        'preset': query.get('preset') or '',
    }

    if has_context_lines:
        filters['contextLines'] = context_lines
    if has_strict_context:
        filters['strictContext'] = strict_context

    return log_service.normalize_filters(filters)


def escape_csv_value(value) -> str:
    """Escape a value for CSV output."""
    if value is None:
        return ''
    s = str(value)
    if ',' in s or '"' in s or '\n' in s:
        return '"' + s.replace('"', '""') + '"'
    return s


def _export_timestamp() -> str:
    """Generate a YYYY.MM.DD_HH-mm-ss timestamp string for file names."""
    now = datetime.now()
    return now.strftime('%Y.%m.%d_%H-%M-%S')


# ---------------------------------------------------------------------------
# App factory (dependency-injection friendly for tests)
# ---------------------------------------------------------------------------

def create_app(folder_path: str, log_service=None) -> FastAPI:
    """
    Create and return a FastAPI application bound to folder_path.
    If log_service is None, a new NEUFLogService is created.
    """
    if log_service is None:
        log_service = NEUFLogService(logger=print)

    app = FastAPI(title='NEUF Log Viewer API', version='1.0.0')

    # CORS — mirrors the JS API
    app.add_middleware(
        CORSMiddleware,
        allow_origins=['*'],
        allow_methods=['*'],
        allow_headers=['*'],
    )

    # -----------------------------------------------------------------------
    # GET /health
    # -----------------------------------------------------------------------

    @app.get('/health')
    async def health():
        scanned = log_service.is_database_scanned(folder_path)
        return JSONResponse({
            'success': True,
            'status': 'healthy',
            'version': '1.0.0',
            'databaseScanned': scanned,
            'folderPath': folder_path,
        })

    # -----------------------------------------------------------------------
    # POST /preset_suggestions
    # -----------------------------------------------------------------------

    @app.post('/preset_suggestions')
    async def preset_suggestions():
        try:
            result = await log_service.get_preset_suggestions(folder_path)
            return JSONResponse(result)
        except Exception:
            return JSONResponse({'success': False, 'error': 'Internal server error'}, status_code=500)

    # -----------------------------------------------------------------------
    # POST /filter_log
    # -----------------------------------------------------------------------

    @app.post('/filter_log')
    async def filter_log(body: FilterLogRequest):
        try:
            # Resolve which filters object to use (new-style or legacy steps)
            raw_filters = body.filters or {}
            if not raw_filters and body.steps:
                raw_filters = (body.steps[0] or {}).get('filters', {})

            parsed_filters = parse_filters_from_request(raw_filters, log_service)
            await log_service.apply_preset(folder_path, parsed_filters)

            result = await log_service.filter_logs(
                folder_path, parsed_filters, {'page': body.page, 'pageSize': body.pageSize}
            )

            # Fetch filter options from the output table
            filter_options_result = await log_service.get_filter_options(
                folder_path, result.get('outputTable')
            )

            if body.raw:
                return JSONResponse({**result, 'filterOptions': filter_options_result.get('data')})

            # Format logs (API layer responsibility)
            formatted_logs = [log_service.format_log_entry(log, 'full') for log in result.get('logs', [])]
            return JSONResponse({
                **result,
                'logs': formatted_logs,
                'filterOptions': filter_options_result.get('data'),
            })

        except Exception:
            return JSONResponse({'success': False, 'error': 'Internal server error'}, status_code=500)

    # -----------------------------------------------------------------------
    # POST /export_log
    # -----------------------------------------------------------------------

    @app.post('/export_log')
    async def export_log(body: ExportLogRequest):
        try:
            raw_filters = body.filters or {}
            if not raw_filters and body.steps:
                raw_filters = (body.steps[0] or {}).get('filters', {})

            fmt = body.format
            if fmt not in ('full', 'compact', 'json', 'csv'):
                fmt = 'full'

            parsed_filters = parse_filters_from_request(raw_filters, log_service)
            await log_service.apply_preset(folder_path, parsed_filters)

            export_page_size = 5000
            first_result = await log_service.filter_logs(
                folder_path, parsed_filters, {'page': 1, 'pageSize': export_page_size}
            )

            # Collect all pages
            if fmt in ('json', 'csv'):
                exported = list(first_result.get('logs', []))
            else:
                exported = [log_service.format_log_entry(log, fmt) for log in first_result.get('logs', [])]

            total_pages = first_result.get('totalPages', 1)
            for page in range(2, total_pages + 1):
                page_result = await log_service.filter_logs(
                    folder_path, parsed_filters, {'page': page, 'pageSize': export_page_size}
                )
                if fmt in ('json', 'csv'):
                    exported.extend(page_result.get('logs', []))
                else:
                    exported.extend([
                        log_service.format_log_entry(log, fmt)
                        for log in page_result.get('logs', [])
                    ])

            ts = _export_timestamp()

            if fmt == 'json':
                filename = f'neuf-logs-export-{ts}-json.json'
                import json as _json
                content = _json.dumps(exported, indent=2, default=str)
                return Response(
                    content=content.encode('utf-8'),
                    media_type='application/json; charset=utf-8',
                    headers={'Content-Disposition': f'attachment; filename="{filename}"'},
                )

            elif fmt == 'csv':
                filename = f'neuf-logs-export-{ts}-csv.csv'
                lines = []
                if exported:
                    headers = ['filename', 'timestamp', 'log_level', 'thread', 'device', 'component', 'message']
                    lines.append(','.join(escape_csv_value(h) for h in headers))
                    for log in exported:
                        row = [
                            f'({log.get("filename", "")})',
                            log.get('timestamp', ''),
                            f'[{log.get("log_level", "")}]',
                            log.get('thread_name', ''),
                            f'<{log.get("device_id", "")}>',
                            f'({log.get("component_name", "")})',
                            log.get('message', ''),
                        ]
                        lines.append(','.join(escape_csv_value(v) for v in row))
                csv_content = '\n'.join(lines)
                return Response(
                    content=csv_content.encode('utf-8'),
                    media_type='text/csv; charset=utf-8',
                    headers={'Content-Disposition': f'attachment; filename="{filename}"'},
                )

            else:
                # 'full' or 'compact' — exported is a list of formatted log objects
                format_suffix = 'compact' if fmt == 'compact' else 'full'
                filename = f'neuf-logs-export-{ts}-{format_suffix}.log'
                lines = [log.get('formattedLog', '') if isinstance(log, dict) else str(log) for log in exported]
                text_content = '\n'.join(lines)
                return Response(
                    content=text_content.encode('utf-8'),
                    media_type='text/plain; charset=utf-8',
                    headers={'Content-Disposition': f'attachment; filename="{filename}"'},
                )

        except Exception:
            return JSONResponse({'success': False, 'error': 'Internal server error'}, status_code=500)

    return app


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

async def _main():
    args = sys.argv[1:]
    if not args:
        sys.stderr.write('❌ Error: Folder path is required\n')
        sys.stderr.write('Usage: python neuf_log_viewer_api.py <folderPath>\n')
        sys.exit(1)

    folder_arg = args[0]
    folder_path = os.path.realpath(os.path.abspath(folder_arg))

    if not os.path.exists(folder_path):
        sys.stderr.write(f'❌ Error: Folder path does not exist: {folder_path}\n')
        sys.exit(1)

    if not os.path.isdir(folder_path):
        sys.stderr.write(f'❌ Error: Path is not a directory: {folder_path}\n')
        sys.exit(1)

    print(f'📁 Folder path: {folder_path}')

    log_service = NEUFLogService(logger=print)
    await log_service.initialize()

    print('\n🔍 Scanning logs...')
    try:
        scan_result = await log_service.scan_logs(folder_path)
    except Exception as scan_err:
        sys.stderr.write(f'\n❌ Failed to scan logs:\n{scan_err}\n')
        sys.exit(1)

    if scan_result.get('success'):
        data = scan_result.get('data', {})
        if not data.get('alreadyScanned'):
            print(f"✅ Scanned {data.get('totalLogs', 0)} log entries from {data.get('filesScanned', 0)} files")
    else:
        sys.stderr.write(f"❌ Failed to scan logs: {scan_result.get('error')}\n")
        sys.exit(1)

    base_port = int(os.environ.get('PORT', '3001'))
    max_attempts = 10

    app = create_app(folder_path=folder_path, log_service=log_service)

    print()
    print('╔════════════════════════════════════════╗')
    print('║   🚀 NEUF Log Viewer API v1.0 (Python) ║')
    print('╚════════════════════════════════════════╝')
    print()
    print(f'🌐 API Server running at http://localhost:{base_port}')
    print(f'📁 Folder path: {folder_path}')
    print()
    print('📋 Available endpoints:')
    print('  POST   /filter_log           - Filter logs (includes filterOptions)')
    print('  POST   /export_log           - Export filtered logs')
    print('  POST   /preset_suggestions   - Get preset filter suggestions')
    print('  GET    /health               - Health check')
    print()
    print('Press Ctrl+C to stop the server')

    # Try successive ports
    for attempt in range(max_attempts):
        port = base_port + attempt
        try:
            config = uvicorn.Config(app, host='0.0.0.0', port=port, log_level='warning')
            server = uvicorn.Server(config)
            await server.serve()
            break
        except OSError:
            if attempt < max_attempts - 1:
                print(f'⚠️  Port {port} is in use, trying {port + 1}...')
            else:
                sys.stderr.write('❌ Could not find an available port.\n')
                sys.exit(1)


if __name__ == '__main__':
    import asyncio
    asyncio.run(_main())
